File: python/src/prts_mcp/data/enemy_stats.py
"""Enemy combat-stats extraction and m_defined override merging.

Extracted from ``data/enemy`` (``extract_enemy_stats``) and ``data/stage_enemy``
(``merge_defined`` / ``stage_specific_enemy_data`` / ``overwritten_enemy_name`` /
``format_stats``) in P3.A. Pure dict logic only — no store/config; callers pass
the shared dataset accessor's result (``levels``) so this module stays
side-effect-free. The TS mirror is ``ts/src/data/enemyStats.ts`` (which also
holds the TS-only Python-formatting shims).
"""
from __future__ import annotations

from typing import Any

from prts_mcp.data.gamedata_attrs import m_value

_IMMUNITY_LABELS: dict[str, str] = {
    "stunImmune": "眩晕",
    "silenceImmune": "沉默",
    "sleepImmune": "睡眠",
    "frozenImmune": "冻结",
    "levitateImmune": "浮空",
    "disarmedCombatImmune": "缴械",
    "fearedImmune": "恐惧",
    "palsyImmune": "瘫痪",
    "attractImmune": "牵引",
}


def extract_enemy_stats(db_entry: dict) -> dict:
    """Extract combat stats from an enemy_database entry into a structured dict.

    Numeric values are pre-formatted as render_enemy_info will emit them
    (e.g. HP with thousands separator), and only non-default fields are
    included.
    """
    attrs: dict = db_entry.get("attributes", {})
    hp = m_value(attrs.get("maxHp"), 0)
    atk = m_value(attrs.get("atk"), 0)
    defense = m_value(attrs.get("def"), 0)
    res = m_value(attrs.get("magicResistance"))
    speed = m_value(attrs.get("moveSpeed"), 0.0)
    atk_time = m_value(attrs.get("baseAttackTime"), 0.0)
    atk_speed = m_value(attrs.get("attackSpeed"), 100.0)
    mass = m_value(attrs.get("massLevel"), 0)
    hp_recovery = m_value(attrs.get("hpRecoveryPerSec"), 0.0)

    immunities = []
    for key, label in _IMMUNITY_LABELS.items():
        if m_value(attrs.get(key), False):
            immunities.append(label)
    lpr = m_value(attrs.get("lifePointReduce"), 0)

    stats: dict[str, Any] = {
        "max_hp": f"{hp:,}" if hp else None,
        "atk": str(atk) if atk else None,
        "def": str(defense) if defense else None,
        "resistance": str(res) if res is not None else None,
        "move_speed": str(speed) if speed else None,
        "attack_interval": f"{atk_time}s" if atk_time else None,
        "attack_speed": str(atk_speed) if atk_speed != 100.0 else None,
        "mass_level": str(mass) if mass else None,
        "hp_recovery_per_sec": str(hp_recovery) if hp_recovery else None,
        "immunities": immunities,
        "life_point_reduce": str(lpr) if lpr else None,
    }

    skills: list[dict] = []
    for s in db_entry.get("skills") or []:
        prefab = s.get("prefabKey", "未知")
        cooldown = s.get("cooldown", "")
        sp_cost = m_value(s.get("spData", {}).get("spCost") if s.get("spData") else None, None)
        init_cd = s.get("initCooldown", "")

        cd_parts = []
        if cooldown:
            cd_parts.append(f"冷却 {cooldown}s")
        if init_cd and init_cd != cooldown:
            cd_parts.append(f"初始 {init_cd}s")
        if sp_cost:
            cd_parts.append(f"SP {sp_cost}")
        timing = "，".join(cd_parts) if cd_parts else ""

        blackboard: list[dict] = s.get("blackboard", [])
        bb_strs = []
        for b in blackboard[:6]:
            key = b.get("key", "")
            val = b.get("value", "")
            if val is not None:
                bb_strs.append(f"{key}={val}")
        bb = "，".join(bb_strs)

        skills.append({"prefab": prefab, "timing": timing, "blackboard": bb})

    stats["skills"] = skills
    return stats


def merge_defined(base: Any, override: Any) -> Any:
    """Merge enemyData dictionaries, applying only m_defined=true overrides."""
    if not isinstance(override, dict):
        return base
    if "m_defined" in override and "m_value" in override:
        return override["m_value"] if override.get("m_defined") else base
    if isinstance(base, dict):
        merged = dict(base)
    else:
        merged = {}
    for key, value in override.items():
        if isinstance(value, dict) and value.get("m_defined") is False:
            continue
        merged[key] = merge_defined(merged.get(key), value)
    return merged


def stage_specific_enemy_data(
    levels: dict[str, dict[int, dict[str, Any]]] | None,
    enemy_id: str,
    level: int,
    overwritten: Any = None,
) -> dict[str, Any] | None:
    """Resolve the effective enemyData for one enemy at one stage level.

    ``levels`` is the raw normalized enemy_database map (the shared dataset
    accessor's result); None/absent enemy falls back to the stage-overwritten
    data when it is dict-shaped.
    """
    db_entry = (levels or {}).get(enemy_id, {})
    base = db_entry.get(level) or db_entry.get(0)
    if base is None:
        return overwritten if isinstance(overwritten, dict) else None
    merged = merge_defined(base, overwritten)
    return merged if isinstance(merged, dict) else base


def overwritten_enemy_name(overwritten: Any) -> str | None:
    if not isinstance(overwritten, dict):
        return None
    name = overwritten.get("name") or overwritten.get("prefabKey")
    return str(m_value(name)) if name else None


def format_stats(enemy_data: dict[str, Any] | None) -> str:
    """One-line compact stat summary for stage-enemies listings."""
    if not enemy_data:
        return "无数据库记录"
    attrs = enemy_data.get("attributes") or {}
    hp = m_value(attrs.get("maxHp"), 0)
    atk = m_value(attrs.get("atk"), 0)
    defense = m_value(attrs.get("def"), 0)
    res = m_value(attrs.get("magicResistance"), 0)
    speed = m_value(attrs.get("moveSpeed"), 0)
    atk_time = m_value(attrs.get("baseAttackTime"), 0)
    parts = [
        f"HP {hp:,}" if isinstance(hp, int) else f"HP {hp}",
        f"ATK {atk}",
        f"DEF {defense}",
        f"RES {res}",
    ]
    if speed:
        parts.append(f"移速 {speed}")
    if atk_time:
        parts.append(f"攻击间隔 {atk_time}s")
    return "；".join(parts)

File: ts/src/data/enemyStats.ts
/**
 * Enemy combat-stats extraction and m_defined override merging.
 *
 * Extracted from data/enemy.ts (extractEnemyStats) and data/stageEnemy.ts
 * (mergeDefined / stageSpecificEnemyData / overwrittenEnemyName /
 * formatStats) in P3.A. Pure dict logic only — no store/config; callers pass
 * the shared dataset accessor's result (levels) so this module stays
 * side-effect-free. Mirrors python/src/prts_mcp/data/enemy_stats.py.
 *
 * NOTE: pythonFloatString / formatNumber / formatFloatLike are TS-only
 * exports with no Python counterpart (Python uses native str()/f-string
 * formatting); they live here because both stat paths consume them.
 */
import { mValue, type MValue } from "./gamedataAttrs.js";

// ---------------------------------------------------------------------------
// enemy_database.json entry shapes (moved from enemy.ts)
// ---------------------------------------------------------------------------

export interface EnemyDbAttrs {
  maxHp?: MValue;
  atk?: MValue;
  def?: MValue;
  magicResistance?: MValue;
  moveSpeed?: MValue;
  baseAttackTime?: MValue;
  attackSpeed?: MValue;
  massLevel?: MValue;
  hpRecoveryPerSec?: MValue;
  spRecoveryPerSec?: MValue;
  lifePointReduce?: MValue;
  stunImmune?: MValue;
  silenceImmune?: MValue;
  sleepImmune?: MValue;
  frozenImmune?: MValue;
  levitateImmune?: MValue;
  disarmedCombatImmune?: MValue;
  fearedImmune?: MValue;
  palsyImmune?: MValue;
  attractImmune?: MValue;
  [key: string]: MValue | undefined;
}

export interface EnemySkill {
  prefabKey?: string;
  priority?: number;
  cooldown?: number;
  initCooldown?: number;
  spData?: { spCost?: MValue };
  blackboard?: Array<{ key?: string; value?: unknown }>;
}

export interface EnemyDbEntry {
  attributes?: EnemyDbAttrs;
  skills?: EnemySkill[] | null;
  talentBlackboard?: Array<{ key?: string; value?: unknown }>;
}

export interface EnemyStatsPayload {
  max_hp: string | null;
  atk: string | null;
  def: string | null;
  resistance: string | null;
  move_speed: string | null;
  attack_interval: string | null;
  attack_speed: string | null;
  mass_level: string | null;
  hp_recovery_per_sec: string | null;
  immunities: string[];
  life_point_reduce: string | null;
  skills: Array<{
    prefab: string;
    timing: string;
    blackboard: string;
  }>;
}

// ---------------------------------------------------------------------------
// TS-only Python-formatting shims
// ---------------------------------------------------------------------------

/** Mimic Python str(float): integers render with a trailing ".0". */
export function pythonFloatString(value: unknown): string {
  if (typeof value === "number" && Number.isInteger(value)) return `${value}.0`;
  return String(value ?? 0);
}

/**
 * Locale-independent thousands separator for ANY number, mirroring Python's
 * unconditional `f"{n:,}"` (the extract path's maxHp formatting).
 */
function withThousands(n: number): string {
  return n.toString().replace(/\B(?=(\d{3})+(?!\d))/g, ",");
}

/**
 * Int-gated thousands separator for unknown values, mirroring Python
 * format_stats's `isinstance(hp, int)` gate (the stage-enemies compact path).
 */
export function formatNumber(value: unknown): string {
  if (typeof value === "number" && Number.isInteger(value)) {
    return withThousands(value);
  }
  return String(value ?? 0);
}

// ---------------------------------------------------------------------------
// Stats extraction (moved from enemy.ts)
// ---------------------------------------------------------------------------

const IMMUNITY_LABELS: Record<string, string> = {
  stunImmune: "眩晕",
  silenceImmune: "沉默",
  sleepImmune: "睡眠",
  frozenImmune: "冻结",
  levitateImmune: "浮空",
  disarmedCombatImmune: "缴械",
  fearedImmune: "恐惧",
  palsyImmune: "瘫痪",
  attractImmune: "牵引",
};

export function extractEnemyStats(dbEntry: EnemyDbEntry): EnemyStatsPayload {
  const attrs = dbEntry.attributes ?? {};
  const hp = mValue<number>(attrs.maxHp, 0) ?? 0;
  const atk = mValue<number>(attrs.atk, 0) ?? 0;
  const def = mValue<number>(attrs.def, 0) ?? 0;
  const res = mValue<number>(attrs.magicResistance);
  const speed = mValue<number>(attrs.moveSpeed, 0) ?? 0;
  const atkTime = mValue<number>(attrs.baseAttackTime, 0) ?? 0;
  const atkSpeed = mValue<number>(attrs.attackSpeed, 100) ?? 100;
  const mass = mValue<number>(attrs.massLevel, 0) ?? 0;
  const hpRec = mValue<number>(attrs.hpRecoveryPerSec, 0) ?? 0;
  const lpr = mValue<number>(attrs.lifePointReduce, 0) ?? 0;

  const immunities: string[] = [];
  for (const [key, label] of Object.entries(IMMUNITY_LABELS)) {
    if (mValue<boolean>(attrs[key], false)) immunities.push(label);
  }

  const skills = (Array.isArray(dbEntry.skills) ? dbEntry.skills : []).map((s) => {
    const prefab = s.prefabKey ?? "未知";
    const cd = s.cooldown;
    const initCd = s.initCooldown;
    const spCost = mValue<number>(s.spData?.spCost);
    const cdParts: string[] = [];
    if (cd) cdParts.push(`冷却 ${cd}s`);
    if (initCd && initCd !== cd) cdParts.push(`初始 ${initCd}s`);
    if (spCost) cdParts.push(`SP ${spCost}`);
    const bbStrs = (Array.isArray(s.blackboard) ? s.blackboard : [])
      .slice(0, 6)
      .filter((b) => b.value != null)
      .map((b) => {
        const value = typeof b.value === "number" ? pythonFloatString(b.value) : b.value;
        return `${b.key ?? ""}=${value}`;
      });
    return {
      prefab,
      timing: cdParts.join("，"),
      blackboard: bbStrs.join("，"),
    };
  });

  return {
    max_hp: hp ? withThousands(hp) : null,
    atk: atk ? String(atk) : null,
    def: def ? String(def) : null,
    resistance: res !== undefined && res !== null ? pythonFloatString(res) : null,
    move_speed: speed ? pythonFloatString(speed) : null,
    attack_interval: atkTime ? `${pythonFloatString(atkTime)}s` : null,
    attack_speed: atkSpeed !== 100 ? pythonFloatString(atkSpeed) : null,
    mass_level: mass ? String(mass) : null,
    hp_recovery_per_sec: hpRec ? pythonFloatString(hpRec) : null,
    immunities,
    life_point_reduce: lpr ? String(lpr) : null,
    skills,
  };
}

// ---------------------------------------------------------------------------
// m_defined override merging (moved from stageEnemy.ts)
// ---------------------------------------------------------------------------

export function mergeDefined(base: unknown, override: unknown): unknown {
  if (!override || typeof override !== "object") return base;
  const overrideRecord = override as Record<string, unknown>;
  if ("m_defined" in overrideRecord && "m_value" in overrideRecord) {
    return overrideRecord["m_defined"] ? overrideRecord["m_value"] : base;
  }
  const merged: Record<string, unknown> =
    base && typeof base === "object" && !Array.isArray(base)
      ? { ...(base as Record<string, unknown>) }
      : {};
  for (const [key, value] of Object.entries(overrideRecord)) {
    if (value && typeof value === "object" && (value as MValue).m_defined === false) continue;
    merged[key] = mergeDefined(merged[key], value);
  }
  return merged;
}

/** Resolve the effective enemyData for one enemy at one stage level. */
export function stageSpecificEnemyData(
  levels: Record<string, Record<number, unknown>> | null | undefined,
  enemyId: string,
  level: number,
  overwritten?: unknown,
): Record<string, unknown> | null {
  const dbEntry = (levels ?? {})[enemyId] ?? {};
  const base = dbEntry[level] ?? dbEntry[0];
  if (!base) return overwritten && typeof overwritten === "object" ? overwritten as Record<string, unknown> : null;
  const merged = mergeDefined(base, overwritten);
  return merged && typeof merged === "object" ? merged as Record<string, unknown> : base as Record<string, unknown>;
}

export function overwrittenEnemyName(overwritten: unknown): string | null {
  if (!overwritten || typeof overwritten !== "object") return null;
  const record = overwritten as Record<string, unknown>;
  const name = record.name ?? record.prefabKey;
  const value = mValue(name);
  return value ? String(value) : null;
}

/** One-line compact stat summary for stage-enemies listings. */
export function formatStats(
  enemyData: { attributes?: EnemyDbAttrs | Record<string, unknown> } | null,
): string {
  if (!enemyData) return "无数据库记录";
  const attrs = (enemyData.attributes ?? {}) as Record<string, unknown>;
  const hp = mValue(attrs.maxHp, 0);
  const atk = mValue(attrs.atk, 0);
  const defense = mValue(attrs.def, 0);
  const res = mValue(attrs.magicResistance, 0);
  const speed = mValue(attrs.moveSpeed, 0);
  const atkTime = mValue(attrs.baseAttackTime, 0);
  const parts = [`HP ${formatNumber(hp)}`, `ATK ${atk}`, `DEF ${defense}`, `RES ${res}`];
  if (speed) parts.push(`移速 ${pythonFloatString(speed)}`);
  if (atkTime) parts.push(`攻击间隔 ${pythonFloatString(atkTime)}s`);
  return parts.join("；");
}
